distributed/recreate_tasks.py

```diff
@@ -0,0 +1,202 @@
+        if isinstance(task, dict):
+            function, args, kwargs = _deserialize(**task)
+            return (function, args, kwargs, deps)
+        else:
+            function, args, kwargs = _deserialize(task=task)
+            return (function, args, kwargs, deps)
+
+    async def _prepare_raw_components(self, raw_components):
+        """
+        Take raw components and resolve future dependencies.
+        """
+        function, args, kwargs, deps = raw_components
+        futures = self.client._graph_to_futures({}, deps)
+        data = await self.client._gather(futures)
+        args = pack_data(args, data)
+        kwargs = pack_data(kwargs, data)
+        return (function, args, kwargs)
+
+    async def _get_components_from_future(self, future):
+        """
+        For a given future return the func, args and kwargs that would be
+        executed remotely. Any args/kwargs that are themselves futures will
+        be resolved to the return value of those futures.
+        """
+        raw_components = await self._get_raw_components_from_future(future)
+        return await self._prepare_raw_components(raw_components)
+
+    def recreate_task_locally(self, future):
+        """
+        For any calculation, whether it succeeded or failed, perform the task
+        locally for debugging.
+
+        This operation should be performed after a future (result of ``gather``,
+        ``compute``, etc) comes back with a status other than "pending". Cases
+        where you might want to debug a successfully completed future could
+        include a calculation that returns an unexpected results. A common
+        debugging process might include running the task locally in debug mode,
+        with `pdb.runcall`.
+
+        Examples
+        --------
+        >>> import pdb                                    # doctest: +SKIP
+        >>> future = c.submit(div, 1, 1)                  # doctest: +SKIP
+        >>> future.status                                 # doctest: +SKIP
+        'finished'
+        >>> pdb.runcall(c.recreate_task_locally, future)  # doctest: +SKIP
+
+        Parameters
+        ----------
+        future : future
+            The same thing as was given to ``gather``.
+
+        Returns
+        -------
+        Any; will return the result of the task future.
+        """
+        func, args, kwargs = sync(
+            self.client.loop, self._get_components_from_future, future
+        )
+        return func(*args, **kwargs)
+
+    async def _get_errored_future(self, future):
+        """
+        For a given future collection, return the first future that raised
+        an error.
+        """
+        await wait(future)
+        futures = [f.key for f in futures_of(future) if f.status == "error"]
+        if not futures:
+            raise ValueError("No errored futures passed")
+        cause = await self.scheduler.get_error_cause(keys=futures)
+        return Future(cause)
```

Comment:
Is this future instatiation necessary? Futures are a bit brittle objects and leak some abstraction. In particular they notify the scheduler about this and perform some reference counting. In particular the reference counting usually needs to be protected by a threading lock, see `Client._refcount_lock`. I'm wondering if this all can be implemented by just passing keys around without messing with ref counting

Reply:
Not necessary. updated to work with they key (or future) instead.